File: src/main/java/features/console/CommandParser.java
package features.console;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static features.console.utils.XmlUtil.asList;

public final class CommandParser {

    static List<Command> commands;

    static {
        try {
            commands = readCommands();
        } catch (Exception e) {
            throw new ConsoleException("Error parsing config file");
        }
    }

    private static List<Command> readCommands() throws ParserConfigurationException, IOException, SAXException {

        File inputFile = new File("commands.xml");
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        Document doc = dBuilder.parse(inputFile);
        doc.getDocumentElement().normalize();

        NodeList commandNodes = doc.getElementsByTagName(Command.COMMAND_TAG);
        var result = new ArrayList<Command>();

        for (var commandNode : asList(commandNodes)){
            result.add(Command.parse(commandNode));
        }

        return result;
    }

    public static void tryExecute(String commandString, CustomPipeline pipeline){
        var args = commandString.split(" ");
        Optional<Command> first = commands.stream().filter(x -> x.name.equals(args[0])).findFirst();

        if(first.isEmpty())
            throw new ConsoleException("unable to find command");

        ConsoleContext ctx = new ConsoleContext(Arrays.copyOfRange(args, 1, args.length), pipeline);
        first.get().execute(ctx);
    }
}

File: src/main/kotlin/com/mirf/features/console/CommandParser.kt
package com.mirf.features.console

import com.mirf.core.pipeline.PipelineBlock
import com.mirf.features.console.utils.asList
import org.xml.sax.SAXException

import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import java.io.File
import java.io.IOException
import java.util.ArrayList

object CommandParser {

    internal var commands: List<Command>

    init {
        try {
            commands = readCommands()
        } catch (e: Exception) {
            throw ConsoleException("Error parsing config file")
        }

    }

    @Throws(ParserConfigurationException::class, IOException::class, SAXException::class)
    private fun readCommands(): List<Command> {

        val inputFile = File("commands.xml")
        val dbFactory = DocumentBuilderFactory.newInstance()
        val dBuilder = dbFactory.newDocumentBuilder()
        val doc = dBuilder.parse(inputFile)
        doc.documentElement.normalize()

        val commandNodes = doc.getElementsByTagName(Command.COMMAND_TAG)
        val result = ArrayList<Command>()

        for (commandNode in commandNodes.asList()) {
            result.add(Command.parse(commandNode))
        }

        return result
    }

    fun tryExecute(commandString: String, pipeline: PipelineBlock<*, *>?) {
        val args = commandString.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val command = commands.first { x -> x.name == args[0] }

        val ctx = ConsoleContext(args, pipeline)
        command.execute(ctx)
    }
}
